Added tests and documentation (#59)
aibaars/dismiss-alerts: README.md

# Dismiss Alerts Action 

> The `dismiss alerts` action closes code scanning alerts based on the `suppression` [property](https://docs.oasis-open.org/sarif/sarif/v2.0/csprd02/sarif-v2.0-csprd02.html#_Toc10127852) in the SARIF file. 

Reviewer: Nit: in code scanning `closed` means `fixed` or `dismissed`, let's use `dismiss` here for clarity.
```suggestion
The `dismiss alerts` action [dismisses](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/managing-code-scanning-alerts-for-your-repository) code scanning alerts based on the `suppression` [property](https://docs.oasis-open.org/sarif/sarif/v2.0/csprd02/sarif-v2.0-csprd02.html#_Toc10127852) in the SARIF file. 
```

There are two required input fields for this action: 
- `sarif-upload-id` - the SARIF identifier
- `sarif-file` - the location of the SARIF file

## Getting Started 

CodeQL populates the `suppression` property in its SARIF output based on the results of `alert-suppression` queries. A user can provide their own custom alert-suppression query, or use the ones that we provide (//lgtm or //codeql style comments).

### Example - CodeQL 

```yaml
name: "Action Test"
on:
 push:
   branches: [main]
 pull_request:
    # The branches below must be a subset of the branches above
   branches: [main]
 workflow_dispatch:

jobs:
  analyze:
    name: Analyze
    runs-on: ubuntu-latest

    strategy:
      fail-fast: false
      matrix:
        language: ["java"]

    steps:
      - name: Checkout repository
        uses: actions/checkout@v2

      - name: Initialize CodeQL
        uses: github/codeql-action/init@v2
        with:
          languages: ${{ matrix.language }}

>           packs: advanced-security-demo/java-alert-suppression

Reviewer: We could use the `//lgtm`-style suppression queries directly from standard query packs. That's maybe easier than having to maintain separate demo packs.

```suggestion
        packs: "codeql/java-queries:AlertSuppression.ql"
```

      - run: |
          javatest/build

      - name: Perform CodeQL Analysis
        id: analyze
        uses: github/codeql-action/analyze@v2
        with:
          output: sarif-results

>       - name: dismiss-alerts
>         uses: ./

Reviewer: Let's add an `if` that checks for the "default" branch, so users get the right behaviour if they copy/paste from the example.

```suggestion
      - name: Dismiss alerts
        if: github.ref == 'main'
        uses: advanced-security/dismiss-alerts
```

        with:
          sarif-id: ${{ steps.analyze.outputs.sarif-id }}
          sarif-file: sarif-results/java.sarif
        env:
          GITHUB_TOKEN: ${{ github.token }}
```

### Third party produced SARIF file 

> The `dismiss alerts` solution can be used with SARIF files from third party providers.

Reviewer: ```suggestion
The `dismiss-alerts` action can be used with SARIF files from third party providers.
```

``` yaml
on:
  push:

jobs:
  check-codeql-versions:
    runs-on: ubuntu-latest

    permissions:
      security-events: write

    steps:
    - name: Checkout code
      uses: actions/checkout@v3
    
    - name: Run SAST scan
      run: sast-scan.sh --output=scan-results.sarif
      
    - name: Upload scan results
      id: upload
      uses: github/codeql-action/upload-sarif
      with:
        sarif_file: scan-results.sarif
        wait-for-processing: true

>     - name: Dismiss alerts
>         uses: advanced-security/dismiss-alerts
>         with:
>           sarif-id: ${{ steps.upload.outputs.sarif-id }}
>           sarif-file: scan-results.sarif
>         env:
>           GITHUB_TOKEN: ${{ github.token }}        

Reviewer: Indentation and `if` condition.

```suggestion
    - name: Dismiss alerts
      if: github.ref == 'main'    
      uses: advanced-security/dismiss-alerts
      with:
        sarif-id: ${{ steps.upload.outputs.sarif-id }}
        sarif-file: scan-results.sarif
      env:
        GITHUB_TOKEN: ${{ github.token }}        
```

```
## High Level Architecture 

The `suppressions[]` object in the sarif is used to create a list of suppressed alerts. The API's are used to retrieve a list of already dismissed alerts. These two lists are mapped using the alert identifier (rule and location).  A comparison is done between these lists and any alert that has not already been dismissed is updated with a PATCH request using the `github/alertUrl` property. The alert `state` is updated to `dismissed` with the `dismissed reason` being `won't fix`.

> ## Features and Limitations 

Reviewer: This is an important section, shall we move it to before the examples?

Author: Good point, moved.

> - This action should run only on the default branch as the dismissal status of an alert is a global property. If this action is run on a push event to a feature branch or pull request then the suppressed alerts will also be dismissed on the default branch. 

Reviewer: This reminded me that we need to add `if` statements to the examples above. Thanks!

> - When an alert is dismissed, a duplicate alert will be marked as fixed. This is because code scanning uses the hash of the alert's line contents as the unique identifier. The suppression comment appears on the same line and therefore a new hash is created, so uniqueness no longer holds and a duplicate is created. Since the original hash no longer exists it is marked autmatically as `fixed`.

Reviewer: This only happens when a new suppression comment is inserted on the same line as the alert. Currently, the `//lgtm` style suppression comments need to be placed at the end of the same line of an alert and therefore have this problem. If we update the LGTM alert-suppression query to also allow `//lgtm` comments to be place on a blank line above the alert then we won't have this problem (in addition it probably looks better too).

```suggestion
- When a suppression comment is added on the line that contains an alert then this alert will be closed and a duplicate alert will be marked as fixed. This is because code scanning uses the hash of the alert's line contents as the unique identifier. The inserted suppression comment changes the contents of the line, and therefore also the hash of the alert. Since the alert hash no longer matches the original alert is considered `fixed` and a new alert is created in its place. The new alert is immediately marked as `dismissed` as a result of the suppression comment. To avoid this problem it is recommended to use a suppression style that allows placing suppression markers on the line before an alert.
```

# Dismiss Alerts 

> The dismiss alerts solution closes code scanning alerts that are commented with a  `codeql` comment in the codebase. This action should only run on the default branch as the dismissal status of an alert is a global property. If this action is run on a push to a feature branch or pull request then the suppressed alerts will also be dismissed on the default branch. 

Reviewer: ```suggestion
The dismiss alerts action closes code scanning alerts that are commented with a `codeql` comment in the codebase. This action should only run on the default branch as the dismissal status of an alert is a global property. If this action is run on a push to a feature branch or pull request then the suppressed alerts will also be dismissed on the default branch. 
```

Author: Updated wording in README.md

> The dismiss alerts solution closes code scanning alerts that are commented with a  `codeql` comment in the codebase. This action should only run on the default branch as the dismissal status of an alert is a global property. If this action is run on a push to a feature branch or pull request then the suppressed alerts will also be dismissed on the default branch. 

Reviewer: The `dismiss-alerts` action is not really tied to CodeQL. The action interprets `suppression` markers in SARIF files to decide which alerts to dismiss. It can be used with SARIF files from third party providers too.

CodeQL includes `suppression` markers in its SARIF output based on the results of `alert-suppression` queries. A user can provide their own custom `alert-suppression` query, or use the ones that we provide (`//lgtm` or `//codeql` style comments).

I think it is important to first explain what the Action does and how it works. Include some example workflow like
```yaml
on:
  push:

jobs:
  check-codeql-versions:
    runs-on: ubuntu-latest

    permissions:
      security-events: write

    steps:
    - name: Checkout code
      uses: actions/checkout@v3
    
    - name: Run SAST scan
      run: sast-scan.sh --output=scan-results.sarif
      
    - name: Upload scan results
      id: upload
      uses: github/codeql-action/upload-sarif
      with:
        sarif_file: scan-results.sarif
        wait-for-processing: true

    - name: Dismiss alerts
        uses: advanced-security/dismiss-alerts
        with:
          sarif-id: ${{ steps.upload.outputs.sarif-id }}
          sarif-file: scan-results.sarif
        env:
          GITHUB_TOKEN: ${{ github.token }}        
```
After that we can explain how the Action can be used together with CodeQL.

Author: Updated wording in README.md

## Example 

```yaml
name: "Example Workflow"
on:
  workflow_dispatch:

jobs:
  analyze:
    name: Analyze
    runs-on: ubuntu-latest

    strategy:
      fail-fast: false
      matrix:
        language: ["java"]

    steps:
      - name: Checkout repository
        uses: actions/checkout@v2

      - name: Initialize CodeQL
        uses: github/codeql-action/init@v2
        with:
          languages: ${{ matrix.language }}
          packs: advanced-security/${{ matrix.language }}-alert-suppression

      - name: Autobuild
        uses: github/codeql-action/autobuild@v2

      - name: Perform CodeQL Analysis
        id: analyze
        uses: github/codeql-action/analyze@v2
        with:
          output: ${{ matrix.language }}.sarif-results

      - name: dismiss-alerts
        uses: ./
        with:
          sarif-id: ${{ steps.analyze.outputs.sarif-id }}
          sarif-file: sarif-results/${{ matrix.language }}.sarif
        env:
          GITHUB_TOKEN: ${{ github.token }}
          
```

## High Level Architecture 

> <img width="609" alt="Screen Shot 2022-11-23 at 11 02 26 pm" src="https://user-images.githubusercontent.com/68650974/203543284-4926768d-e1d1-49f1-baba-b04a978b0e2d.png">

Reviewer: ```suggestion
<img width="609" alt="architecture diagram" src="https://user-images.githubusercontent.com/68650974/203543284-4926768d-e1d1-49f1-baba-b04a978b0e2d.png">
```

Reviewer: I'm not entirely sure we need to include the picture in the readme. It might be too detailed for end-users.

Author: Agreed, removed the diagram.
